### ISSUE TITLE
Plugins exposing multiple rules

### ISSUE THREAD

**Issue by dryoma** [stylelint/stylelint#1174]
Is there a way to implement functionality that would allow one plugin to expose multiple rules?

Right now it is "one plugin = one rule", so if more than one rule is implemented in a repo, then 1) the repo has to create separate plugins for each of the rules, and 2) including them one by one:

```json
.stylelintrc.json:
...
  "plugins": [
    "repo-name/src/rules/rule-name-1",
    "repo-name/src/rules/rule-name-2",
    "repo-name/src/rules/rule-name-3"
  ],
  "rules": {
    "rule-name-1": true,
    "rule-name-2": false,
    "rule-name-3": 1,
    ...
  }
...
```

What if a plugin could work like this:

```json
...
  "plugins": [
    "repo-name"
  ],
  "rules": {
    "rule-name-1": true,
    "rule-name-2": false,
    "rule-name-3": 1,
    ...
  }
...
```
Less repetition and, I think, an interface for multiple rules would be simpler (one `createPlugin` instead of multiple calls in multiple files).

**Comment by evilebottnawi** [comment]
@dryoma default all rules are disable, i think it is not good solution

**Comment by jeddy3** [comment]
I'm not sure @dryoma is suggesting enabling the rules by default. Just *bundling them* into one plugin i.e. if you specify a plugin bundle and don't configure any rules from it, then it'll do nothing.

@dryoma, @davidtheclark wrote the plugin system and so he's probably the best person to comment on the feasibility of bundling rules into a single plugin.

**Comment by dryoma** [comment]
> Just bundling them into one plugin i.e. if you specify a plugin bundle and don't configure any rules from it, then it'll do nothing.

That's right, just bundling a bunch of rules in one plugin.

**Comment by davidtheclark** [comment]
Have you tried it? I think you probably can't now but it should be an easy feature to add. At a glance I'd guess you need to modify some logic around https://github.com/stylelint/stylelint/blob/master/src/postcssPlugin.js#L51 to allow for modules that export an array of plugin objects instead of a single one. And of course write some tests.

**Comment by mehcode** [comment]
@davidtheclark  / @jeddy3 

Before this gets released I want to throw in a suggestion. Can we require plugin namespaces when configuring rules?

```
...
  "plugins": [
    "repo-name"
  ],
  "rules": {
    "rule-name-1": true,
    "repo-name/rule-name-1": true,
    "repo-name/rule-name-2": false,
    "repo-name/rule-name-3": 1,
    ...
  }
...
```

Otherwise adding any rule to stylelint is technically an API break-age (major version bump) as their is an implied API (in plugins) to add rules of any name that is not in stylelint.

**Comment by jeddy3** [comment]
> Otherwise adding any rule to stylelint is technically an API break-age (major version bump) as there is an implied API (in plugins) to add rules of any name that is not in stylelint.

I'm going to reopen this issue as this is a very valid concern.

I think we're got an opportunity here to get this right now -- as there is only a handful of plugins in the wild and `stylelint-plugin-scss` (the first plugin to include a set of rules) isn't out yet.

So, do we require all plugins rules configs to be namespaced? It would certainly make sense for the likes of `stylelint-plugin-scss`. However, it would be quite clunky for plugins which are single rules e.g. `stylelint-selector-bem-pattern`.

It feels like we're approaching ground already covered by PostCSS's single purpose plugins, PostCSS plugin packs and how eslint handles its plugins.

Anyone got any suggestions on the best way forward here?

**Comment by mehcode** [comment]
:+1: To requiring _some kind of_ namespace for all rules regardless of rule count.

---

Just to throw an idea out there. We already special case single-rule plugins (export object vs array of objects). We could special case the rule syntax as well here.

```yaml
{
  "plugins": [ 
    "selector-max-depth",
    "selector-bem-pattern",
    "scss",
  ],
  "rules": {
    "scss/extend-require-placeholder": true,
    "scss/selector-no-redundant-parent": true,
    "selector-max-depth/": 2,
    "selector-bem-pattern/": {
      # ...
    }
  }
}
```

- We could chose a symbol other than `/` that might look better trailing off like that (`scss#rule` or `selector-bem-pattern#` / `scss@rule` or `selector-bem-pattern@` )

- I would lean towards requiring `stylelint-plugin-` in the npm name as eslint does. It'd shorten the plugin array and the namespace names.

**Comment by dryoma** [comment]
Just a thought: how about a plugin having a namespace as an independent parameter, not necessarily equal to the plugin name? Say, some `stylelint-do-something-crazy` plugin can set a namespace as `crazy`. Although it might cause collisions if multiple plugins set some common thing (`scss`) as a namespace.

**Comment by kristerkari** [comment]
As @mehcode already suggested, the ESLint way is a good option:

For example:
[eslint-plugin-react](https://github.com/yannickcr/eslint-plugin-react)


```json
{
  "plugins": [
    "react"
  ],
  "rules": {
    "react/forbid-prop-types": 1,
    "react/jsx-boolean-value": 2,
    "react/jsx-closing-bracket-location": 2
  }
}
```

**Comment by davidtheclark** [comment]
Good suggestion!

I think we don't lose anything by making the behavior consistent whether the rule is from a single-rule plugin or a multi-rule plugin. So we should just make it consistent.

To avoid a breaking change on stylelint's side we could just issue a deprecation warning if a plugin ruleName does not have a prefix.

I'm fine with using `/` as the prefix punctuation — because ESLint already uses that, so why not? 

To summarize, here's what I'm proposing:

We simply add a deprecation warning whenever a plugin rule name does not include a `/`, and document this expectation. I don't think we need to place any particular requirements on what prefix names need to be or anything like that. Simply ensuring that it has a `/` will ensure that it will not clash with stylelint core rules.

Sound good? Or objections?

**Comment by jeddy3** [comment]
> Simply ensuring that it has a / will ensure that it will not clash with stylelint core rules.

It does seem to be the simplest solution to addressing the original "API break-age" issue.

> I don't think we need to place any particular requirements on what prefix names need to be or anything like that. 

Do we want to document some recommendations though to help bring some consistency to plugins? e.g. 

We recommended that :
- Rule names within a multi-rule plugin should be prefixed with a common name e.g. `scss-like/dollar-variable-pattern`
- The rule name within a single-rule plugin can be prefix-less, but must begin with a `/` to distinguish them from stylelint core rules e.g. `/selector-bem-pattern`

Is that what we're doing here, or have I gotten the wrong end of the stick?

**Comment by dryoma** [comment]
Great, it works. Thanks!
